### ISSUE TITLE
Apps with sharedUserIds and the app internet permission label

### ISSUE THREAD

**Issue by dragonGod97677** [d4rken/sdmaid-public#5607]
As i said on #5591

I think even an app is doesn't have a ``android.permission.INTERNET`` but have a ``android:sharedUserId`` can still connect to internet if the original app with the same user id have internet permission 
And system apps with ``android:sharedUserId="android.uid.system"`` but having a no internet permission tag should be excluded.

**Comment by TheMysticS** [comment]
Can you explain that in layman terms?

Are you saying that a developer can have 5 apps, out of which 4 do not have internet permission on their own, but the 5th one does. And of all these apps have a shared user ID, those 4 without internet permission can still access internet using the internet permissions of the 5th one?

That is quite interesting. Do these apps have to be from a same developer or can that also be manipulated?

**Comment by dragonGod97677** [comment]
> Can you explain that in layman terms?
> 
> Are you saying that a developer can have 5 apps, out of which 4 do not have internet permission on their own, but the 5th one does. And if all these apps have a shared user ID, those 4 without internet permission can still access internet using the internet permissions of the 5th one?

Yes

> That is quite interesting. Do these apps have to be from a same developer or can that also be manipulated?

I think it cannot be manipulated because of apk signature

**Comment by dragonGod97677** [comment]
I learnt it from NetGuard app

Try it by blocking a system app that have a same shared id but doesn't have internet permission

**Comment by TheMysticS** [comment]
@d4rken 

I have an idea for an app? Where do I post it? Also, are you open to develop a new app in addition to SD Maid?

**Comment by TheMysticS** [comment]
Is it because of apk signature or is it because it is dependent on something else? Asking because even if you have root, you won't be able to replace the stock app with the other one. Example would be removing the file manager app on a OnePlus device and replacing it with the one from Huawei. I think there is more to it than just signature.

**Comment by dragonGod97677** [comment]
![Screenshot_2022-07-13-17-38-03-45.png](https://user-images.githubusercontent.com/38875906/178702678-800469b6-14a3-4cdd-a339-cdba8e066fd9.png)

**Comment by dragonGod97677** [comment]
![Screenshot_2022-07-13-17-39-48-38.png](https://user-images.githubusercontent.com/38875906/178703157-e940c009-4dd5-403f-b827-ea81acafa274.png) http://developer.android.com/guide/topics/manifest/manifest-element.html last screenshot i gonna send about it

**Comment by dragonGod97677** [comment]
> Is it because of apk signature or is it because it is dependent on something else? Asking because even if you have root, you won't be able to replace the stock app with the other one. Example would be removing the file manager app on a OnePlus device and replacing it with the one from Huawei. I think there is more to it than just signature.

No, not replacing a app

**Comment by dragonGod97677** [comment]
![IMG20220713175801.jpg](https://user-images.githubusercontent.com/38875906/178707071-3e81eb64-2fbc-43d6-a962-6d0323808951.jpg)

**Comment by d4rken** [comment]
Checking for shared IDs etc. can make this a bit more accurate, but doesn't cover everything.
A dev can write a mechanism (files, content-provider, broadcasts) that allows two complete different apps to exchange data.
This is infeasible for SD Maid to detect.

Removing the tag from system apps is no issue.

Not sure if detecting sharedIds for user apps is worth it, as it's not necessary for them to exchange data...


> I have an idea for an app? Where do I post it? Also, are you open to develop a new app in addition to SD Maid?

Email or Discord #offtopic maybe?

**Comment by TheMysticS** [comment]
> @TheMysticS ![Screenshot_2022-07-13-17-35-35-73.png](https://user-images.githubusercontent.com/38875906/178702222-40743c54-a3f5-4671-8f65-223d6c4b3c88.png)

On your 1st screenshot, there is mention of 'Certificate" and not signature. This is what I asked you earlier. Am I right in thinking that signature is a mechanism only to check if an app is tampered with or not?

User ID and Certificates are mechanisms to manage permissions and access to data. Am I right?

**Comment by TheMysticS** [comment]
> Checking for shared IDs etc. can make this a bit more accurate, but doesn't cover everything. A dev can write a mechanism (files, content-provider, broadcasts) that allows two complete different apps to exchange data. This is infeasible for SD Maid to detect.
> 
> Removing the tag from system apps is no issue.
> 
> Not sure if detecting sharedIds for user apps is worth it, as it's not necessary for them to exchange data...
> 
Is it possible to add 'SharedUserID' tag/ label to apps? If not on the main screen itself, atleast on the app page that we see when clicking on an app?

> > I have an idea for an app? Where do I post it? Also, are you open to develop a new app in addition to SD Maid?
> 
> Email or Discord #offtopic maybe?

Sure. I'll post it shortly under #Off-Topic.

**Comment by d4rken** [comment]
> Is it possible to add 'SharedUserID' tag/ label to apps? If not on the main screen itself, atleast on the app page that we see when clicking on an app?

Will that be useful to most users? :thinking:

**Comment by TheMysticS** [comment]
> > Is it possible to add 'SharedUserID' tag/ label to apps? If not on the main screen itself, atleast on the app page that we see when clicking on an app?
> 
> Will that be useful to most users? 🤔

It will be useful to those who care and want to have greater control over their apps. This is something new that I learnt only today (an app without internet access can still access internet through other means).

**Comment by d4rken** [comment]
Let me rephrase: For most users of SD Maid.  Any feature we add increases maintenance effort. 😔

This is on the edge of being in scope.

**Comment by TheMysticS** [comment]
> Let me rephrase: For most users of SD Maid. Any feature we add increases maintenance effort. 😔
> 
> This is on the edge of being in scope.

Since an app without internet permissions can still access internet using SharedUserID or other means, the 'No internet permissions' label doesn't guarantee complete lack of internet access, unless the app also does not have a SharedUserID status or other ways of accessing internet.

Since you have (thankfully) agreed to add this new label, you might as well make it a more accurate representation of reality. In this context, here is another suggestion:

An app with no internet permissions, as well as no SharedUserID status (or other means of accessing internet), will have a green background.

An app with no internet permissions but with SharedUserID status will have a different coloured background, such as orange, so the user becomes aware of it on the main screen itself.

Also, clicking on the app would additionally also show the list of all other apps with the same SharedUserID status.

**Comment by dragonGod97677** [comment]
> > @TheMysticS ![Screenshot_2022-07-13-17-35-35-73.png](https://user-images.githubusercontent.com/38875906/178702222-40743c54-a3f5-4671-8f65-223d6c4b3c88.png)
> 
> On your 1st screenshot, there is mention of 'Certificate" and not signature. This is what I asked you earlier. Am I right in thinking that signature is a mechanism only to check if an app is tampered with or not?
> 
> User ID and Certificates are mechanisms to manage permissions and access to data. Am I right?

First, well, i dont make apps, i learn these via apps, reading app's AndroidManifest.xml file and with research. Also to answer your question, yes i guess.

I only raised this issue because yeah, app can still have internet through shared id

**Comment by TheMysticS** [comment]
> > > @TheMysticS ![Screenshot_2022-07-13-17-35-35-73.png](https://user-images.githubusercontent.com/38875906/178702222-40743c54-a3f5-4671-8f65-223d6c4b3c88.png)
> > 
> > 
> > On your 1st screenshot, there is mention of 'Certificate" and not signature. This is what I asked you earlier. Am I right in thinking that signature is a mechanism only to check if an app is tampered with or not?
> > User ID and Certificates are mechanisms to manage permissions and access to data. Am I right?
> 
> First, well, i dont make apps, i learn these via apps, reading app's AndroidManifest.xml file and with research. Also to answer your question, yes i guess.
> 
> I only raised this issue because yeah, app can still have internet through shared id, i find about it because i used NetGuard before and notified me about blocking app with shared id and the app has Internet Permission and i just want the No Internet Permission tag be accurate

Glad that you brought it up. I learnt something very important and useful. :)

**Comment by dragonGod97677** [comment]
Though im been thinking, why not just name the tag Offline App and apply the tag if

It has no internet permission
It has no shared tag
   - if it has a shared tag, and also if the shared tag is named like a package name, lookup for that app and check if it has Internet Permission, if the origin app doesn't have the internet permission, then apply the offline tag, otherwise dont apply.
   - if the shared id is ``android.uid.system``, then dont apply the tag

**Comment by dragonGod97677** [comment]
Using No internet permission is just checks the app if it doesn't have internet permission

**Comment by TheMysticS** [comment]
> Using No internet permission is just checks the app if it doesn't have internet permission

Did you see this comment:

https://github.com/d4rken/sdmaid-public/issues/5607#issuecomment-1183471922

**Comment by dragonGod97677** [comment]
> > Using No internet permission is just checks the app if it doesn't have internet permission
> 
> Did you see this comment:
> 
> [#5607 (comment)](https://github.com/d4rken/sdmaid-public/issues/5607#issuecomment-1183471922)
There will be new tag? Alright.

**Comment by TheMysticS** [comment]
> > > Using No internet permission is just checks the app if it doesn't have internet permission
> > 
> > 
> > Did you see this comment:
> > [#5607 (comment)](https://github.com/d4rken/sdmaid-public/issues/5607#issuecomment-1183471922)
> 
> There will be new tag? Alright.

Not new tag, but same tag with a different background colour to differentiate between the two.

**Comment by dragonGod97677** [comment]
> > > > Using No internet permission is just checks the app if it doesn't have internet permission
> > > 
> > > 
> > > Did you see this comment:
> > > [#5607 (comment)](https://github.com/d4rken/sdmaid-public/issues/5607#issuecomment-1183471922)
> > 
> > 
> > There will be new tag? Alright.
> 
> Not new tag, but same tag with a different background colour to differentiate between the two.

Ahh, alright, i get it but i hope they would really fully read the documentation about it when this get on sd maid and they found out why there's different colors of no internet permission tag

**Comment by d4rken** [comment]
Example, some apps with `sharedUserId`:

```java
PKG: com.google.ar.core SharedId: com.google.ar.core
PKG: com.google.android.cellbroadcastservice SharedId: android.uid.networkstack
PKG: com.android.providers.telephony SharedId: android.uid.phone
PKG: com.android.mtp SharedId: android.media
PKG: com.google.android.networkstack.tethering SharedId: android.uid.networkstack
PKG: com.qualcomm.qcrilmsgtunnel SharedId: android.uid.phone
PKG: com.android.shell SharedId: android.uid.shell
PKG: com.android.providers.userdictionary SharedId: android.uid.shared
PKG: com.android.keychain SharedId: android.uid.system
PKG: com.android.soundpicker SharedId: android.media
PKG: com.android.service.ims.presence SharedId: android.uid.phone
PKG: com.google.RilConfigService SharedId: android.uid.phone
PKG: com.microsoft.office.officehubrow SharedId: com.microsoft.office
PKG: com.google.android.marvin.talkback SharedId: com.google.android.marvin.feedback
PKG: com.android.dynsystem SharedId: android.uid.system
PKG: com.qualcomm.uimremoteclient SharedId: android.uid.phone
PKG: com.google.android.gsf SharedId: com.google.uid.shared
PKG: com.google.android.apps.maps SharedId: com.google.android.apps.maps
PKG: com.google.android.networkstack SharedId: android.uid.networkstack
PKG: com.qti.xdivert SharedId: android.uid.qtiphone
PKG: com.android.settings SharedId: android.uid.system
PKG: com.android.providers.calendar SharedId: android.uid.calendar
PKG: com.android.server.telecom SharedId: android.uid.system
PKG: com.android.providers.downloads SharedId: android.media
PKG: com.android.providers.downloads.ui SharedId: android.media
PKG: com.android.providers.blockednumber SharedId: android.uid.shared
PKG: com.qualcomm.qti.cne SharedId: com.qualcomm.qti.qtidataservices
PKG: com.google.SSRestartDetector SharedId: android.uid.system
PKG: com.mxtech.videoplayer.ad SharedId: com.mxtech
PKG: com.google.android.networkstack.permissionconfig SharedId: android.uid.networkstack
PKG: com.android.nfc SharedId: android.uid.nfc
PKG: com.android.emergency SharedId: com.android.emergency.uid
PKG: com.android.location.fused SharedId: android.uid.system
PKG: de.myaudi.mobile.assistant SharedId: com.audi.myaudi.userid
PKG: com.android.localtransport SharedId: android.uid.system
PKG: com.android.sdm.plugins.sprintdm SharedId: android.uid.phone
PKG: com.maxmpz.audioplayer.unlock SharedId: com.maxmpz
PKG: vendor.qti.iwlan SharedId: com.qualcomm.qti.qtidataservices
PKG: android SharedId: android.uid.system
PKG: com.google.android.gms SharedId: com.google.uid.shared
PKG: com.android.omadm.service SharedId: com.android.uid.omadm.service
PKG: com.google.android.projection.gearhead SharedId: com.google.android.projection.gearhead
PKG: com.android.se SharedId: android.uid.se
PKG: com.android.bluetooth SharedId: android.uid.bluetooth
PKG: com.android.ons SharedId: android.uid.phone
PKG: vendor.qti.hardware.cacert.server SharedId: com.qualcomm.qti.qtidataservices
PKG: com.termux SharedId: com.termux
PKG: com.android.providers.contacts SharedId: android.uid.shared
PKG: com.android.service.ims SharedId: android.uid.phone
PKG: com.maxmpz.audioplayer SharedId: com.maxmpz
PKG: org.mozilla.firefox SharedId: org.mozilla.firefox.sharedID
PKG: com.android.providers.settings SharedId: android.uid.system
PKG: com.qti.qualcomm.datastatusnotification SharedId: android.uid.phone
PKG: com.android.systemui SharedId: android.uid.systemui
PKG: com.android.phone SharedId: android.uid.phone
PKG: com.android.mms.service SharedId: android.uid.phone
PKG: com.qualcomm.uimremoteserver SharedId: android.uid.phone
PKG: com.android.inputdevices SharedId: android.uid.system
PKG: com.android.calllogbackup SharedId: android.uid.shared
PKG: com.android.wallpaperbackup SharedId: android.uid.system
PKG: com.android.providers.media SharedId: android.media
PKG: com.android.stk SharedId: android.uid.phone
PKG: com.qualcomm.qti.uceShimService SharedId: android.uid.phone
PKG: com.google.android.calendar SharedId: com.google.android.calendar.uid.shared
```

**Comment by TheMysticS** [comment]
Let me rephrase this.

If an app has no internet permissions of its own, but has a SharedUserID:

1. If all other apps that use the same ID do not also have internet permissions, then the 'No Internet' label with a green background can be applied.

2. If any of the other apps that uses the same ID has internet permission, then the 'No Internet' label will be applied with a different coloured background.

3. If an app has no internet permissions but uses a SharedUserID, but if any app with the same ID and having internet permission is NOT installed, this app will still have a 'No Internet' label with a green background, until such time that any of the other apps with the same ID and having internet permissions is installed.

In summary,

Using a SharedUserID alone will not make a difference unless another app with the same ID is installed and has internet permissions.

**Comment by d4rken** [comment]
Searching the list using the sharedUserId is now also possible.

**Comment by TheMysticS** [comment]
> Before ![Screenshot from 2022-07-15 13-30-55](https://user-images.githubusercontent.com/1439229/179215053-a41e1eb3-f436-410a-8114-554ec41a11c4.png)
> 
> After ![Screenshot from 2022-07-15 14-02-54](https://user-images.githubusercontent.com/1439229/179219421-12b07f8c-7929-420a-b6f0-af8fb1b00d44.png)
> 
Let's assume the PowerAmp Unlocker app does NOT have internet permission, but the main player app has it.

If user installs only the Unlocker app, SD Maid would show the 'No Internet' label in Green.

If the user installs the main app too, SD Maid will now change the colour of that label to Orange, like in your screenshots.

So if an app without internet permission of its own has other means of using internet (be it via apps with the same SharedUserID or using other means), then it's 'No Internet' label will have an Orange background.

**Can you also add the details of which apps or processes that such apps depend on for potential internet access?** These details should show up if the user clicks on the app (the screen where SD Maid shows all the app details, along with the ability to uninstall it).

**Comment by d4rken** [comment]
> Can you also add the details of which apps or processes that such apps depend on for potential internet access? These details should show up if the user clicks on the app (the screen where SD Maid shows all the app details, along with the ability to uninstall it).

Yes. Please provide feedback after the next update. :+1: 

> which apps or processes 

FYI, the lowest common denominator at which SD Maid works is "package". Normal apps, libraries, instant apps, are all still packages. Other processes, e.g. from system binaries are out of scope for SD Maid.

**Comment by d4rken** [comment]
> Can you also add the details of which apps or processes that such apps depend on for potential internet access? These details should show up if the user clicks on the app (the screen where SD Maid shows all the app details, along with the ability to uninstall it).

![Screenshot from 2022-07-15 14-42-15](https://user-images.githubusercontent.com/1439229/179225126-0f5d8d0c-3dd8-4c05-b731-8103ad8a7c5a.png)

**Comment by TheMysticS** [comment]
> Other processes, e.g. from system binaries are out of scope for SD Maid.
>
If an app uses a system component for internet access, then it will anyway not have the 'No Internet' label. I guess such apps would be mostly those from the OEM and Google, and not any 3rd party app (please correct me if I am wrong).
